contracts/src/gateway/router/CheckpointingFacet.sol

```diff
@@ -52,9 +52,42 @@ contract CheckpointingFacet is GatewayActorModifiers {
         bytes32 membershipRootHash,
         uint256 membershipWeight
     ) external systemActorOnly {
-        if (checkpoint.blockHeight % s.bottomUpCheckPeriod != 0) {
-            revert InvalidCheckpointEpoch();
-        }
+        // TODO: Discussion:
+        //
+        // This check is removed:
+        //
+        // if (checkpoint.blockHeight % s.bottomUpCheckPeriod != 0) {
+        //    revert InvalidCheckpointEpoch();
+        // }
+        //
+        // First of all, the check is buggy in the first place as it does not support
+        // a full message batch before a checkpoint period is reached.
+        // Then, `prevBottomUpCheckpointHeight` is introduced, as long as BU checkpoints
+        // can form a chain, it should be fine.
+        //
+        // Instead, one can add the following check:
+        //
+        // if (checkpoint.prevHeight != s.prevBottomUpCheckpointHeight) {
+        //     revert InvalidCheckpointEpoch();
+        // }
+        //
+        // However, we are restricting to ONLY sequential creation
+        // of bottom up checkpoints. One can only submit the next one if the current
+        // checkpoint quorum is reached. This would slow down the creation process.
+        //
+        // Or we can enforce:
+        //
+        // checkpoint.msgs.length != 0 || s.prevConfigNumber < checkpoint.nextConfigurationNumber
+        //
+        // so that no empty checkpoint will be created. The drawback is we might have long block intervals with
+        // no checkpoints at all. We can get around this with some `maxBUGap` or something along this line, so that
+        // when no bottom up checkpoints, we dont allow long gap where there is no bottom up checkpoints. This also
+        // has sequential submission restriction
+        //
+        // Or we dont enforce any check at all here, so that we let the child subnet decide whatever checkpoint
+        // the validator wants to create, but as long as quorum is reached and submitted to the parent, the parent
+        // will enforce checkpoints are sequential. This initial implementation took this approach.
```

Comment:
But the safest approach is enforcing:
```
checkpoint.prevHeight != s.prevBottomUpCheckpointHeight
```
If somehow the child decided to the `checkpoint.prevHeight` is not the same as that in the parent, then, the checkpointing is blocked.

Reply:
Do you mean _reverting_ if that condition is true, or enforcing that it _is_ true? 

I don't understand the "But". You would the parent enforce sequential checkpoints without this check?

Reply:
I am slightly confused by the comment, but to be clear I think sequential checkpoints should be enforced by the checks here, we should not admit out-of-order checkpoints.

Reply:
Ah, okay, I see this method is running on the child subnet :upside_down_face: